### DISCUSSION TITLE
need help cleaning up software listing image

### DISCUSSION THREAD

**Discussion by kenrector** [ImageMagick/ImageMagick#1916] [General] [answered]
I'm preparing a couple hundred .png images of software listings printed by line printer on lined paper.  Could someone suggest a method to remove the horizontal lines occurring every 1/2 inch (12mm) or so down the page.  I'm using ocropus for the ocr.  Here is a sample page.
![image101](https://user-images.githubusercontent.com/9969282/79912537-072fc580-83d7-11ea-90b7-7a785069f49d.png)

**Comment by fmw42** [comment, marked as answer]
Perhaps you should post the original image. There may be better ways to convert it to keep better text quality. Your posted image probably will not OCR well as the quality of the characters is horrible.

But you can remove horizontal lines use -morpholgoy close with a horizontal rectangular kernel that is longer than the width of the widest character

**Reply by kenrector** [reply]
Fred,
The horrible quality in the posted image is from a png produced by converting the pdf to png using the following command. 

convert "sym.pdf[175]" -set units PixelsPerInch -density 300 image175.png

I set the density to 300 as that is what OCRopus favored.  But, I don't think I understand that very well.  The conversions below using preview result in 300 ppl by default.

Concerning the original image.  Below are screen shots of images produced from the pdf using MAC preview instead of convert.  The uppermost one is saved as .png and the lower as .jpg.  Preview doesn't seem to offer any parameters for adjusting the quality so these are as is.  To me, these look pretty much the same except the jpg is fuzzier but has the same missing pixels as the png.  Am I right to use the png?  Is there some way to fill in missing info?  I kind of doubt that.  I'll experiment with both conversions when I start getting good results from ocropus.

![Screen Shot 2020-04-21 at 5 26 34 PM](https://user-images.githubusercontent.com/9969282/79927901-11ad8780-83f6-11ea-8712-588130506238.png)
![Screen Shot 2020-04-21 at 5 27 34 PM](https://user-images.githubusercontent.com/9969282/79927923-29850b80-83f6-11ea-97fd-a9e8e1c05ca1.png)

Here's a sample from the ocropus ocr using the better quality png conversion from preview,  Of course not everything is this good, but it's getting better.

![Screen Shot 2020-04-21 at 5 39 23 PM](https://user-images.githubusercontent.com/9969282/79928878-09a31700-83f9-11ea-9730-1cbe7992b6ab.png)
![Screen Shot 2020-04-21 at 5 39 37 PM](https://user-images.githubusercontent.com/9969282/79928893-10ca2500-83f9-11ea-8f1d-0be12c503fa1.png)

I haven't had a chance to try the -morphology feature yet.
Thanks for your help

**Comment by snibgo** [comment]
I suppose your PDF contains embedded raster images, probably one image per page. If so, then I suggest you use "pdfimages" to extract the raster images from the PDF, not ImageMagick.

pdfimages will create image files directly from the embedded PDF images. If you use IM, then you have to worry about "-density", and resampling which will happen unless you get the correct density.

When you have done that, please post a link to one of the extracted images.

**Comment by kenrector** [comment]
I tried to use fred's morphology 
<http://www.fmwconcepts.com/imagemagick/morphology/index.php> script but 
got the following error:   I changed the script to echo $1 and $infile 
and they both show the filename properly.   I can run convert with that 
image file just fine.  I recently upgraded to OSX Ctatalina which uses 
zsh and I changed the first line of the script to #!/bin/zsh to no 
avail.  I wonder what could be wrong?



sds/ocropus_symbol/morph % ../morphology   -t close image-000.png result.png
dyld: Library not loaded: 
/ImageMagick-7.0.10/lib/libMagickCore-7.Q16HDRI.7.dylib
   Referenced from: /Users/admin/applications/ImageMagick-7.0.10/bin/convert
   Reason: image not found
../morphology: line 312:  8112 Abort trap: 6           convert -quiet 
"$infile" +repage "$TMP"

--- FILE image-000.png NOT READABLE OR HAS ZERO SIZE ---


sds/ocropus_symbol/morph %         ls -l image.png
-rw-r--r--@ 1 admin  staff  12715 Apr 22 22:25 image.png


Ken

snibgo wrote on 4/21/20 7:13 PM:

**Reply by fmw42** [reply]
Use Imagemagick morphology. See https://imagemagick.org/Usage/morphology/

**Reply by kenrector** [reply]
I did use pdfimages and the images are much cleaner.  Thanks for the hint.
Moving on I used:
convert image.png -morphology close rectangle:40x1 result.png
composite result.png image.png -compose dst_out new.png  

image.png
![image](https://user-images.githubusercontent.com/9969282/80131698-76c8c080-854f-11ea-8a47-b48c20865e24.png)

result.png
![result](https://user-images.githubusercontent.com/9969282/80131717-7d573800-854f-11ea-8bbc-2b74f11edda8.png)

the composite  command  produced new.png 
![new](https://user-images.githubusercontent.com/9969282/80131803-a37cd800-854f-11ea-8eea-e0a50b2c6633.png)

Changing the order of files in the composite command made no difference.
Identify of the files gave this.
image.png PNG 1760x2673 1760x2673+0+0 8-bit Gray 256c 117849B 0.000u 00.000
result.png PNG 1760x2673 1760x2673+0+0 8-bit Gray 2c 12394B 0.000u 0:00.001

Is there something I need to do about transparency?  It seems that composite is what's needed to remove the lines, but maybe not?  What am I missing?

**Reply by fmw42** [reply]
Try the following. You can try adjusting as desired:

`convert image.png \( -clone 0 -morphology close rectangle:40x1 -negate \) -compose plus -composite result.png`

**Reply by kenrector** [reply]
So, what I needed was the -negate and compose plus, instead of dst_out.

With that I have something close, but larger rectangle heights, like 40x2 or 40x4 give broader results instead of narrower.  It seems I need 40x.5.  ;-).  

result.png with 40x1
![resultx1](https://user-images.githubusercontent.com/9969282/80161011-ee184780-8583-11ea-9e89-f59f31e9a142.png)

thanks for your help
